### PR TITLE
[gfx10] Winograd f3x2 v21_1_2 update

### PR DESCRIPTION
**Pull request by Slimakanzer**
This PR adds _Winograd v21_1_2 f3x2_ shaders and upgrades Winograd f3x2 solver.

## Shader changes compared with _v16_5_0_
- Supports extended cases set
- Supports WrW direction
- Supports grouped convolution
- Tensor offset support (64 bit byte offsets)

⚠️ fp16 kernels will be added in a subsequent PR

### CODE REVIEW

File: src/solver/conv_winoRxS_f3x2.cpp [hunk 1]
```diff
@@ -32,154 +32,42 @@
...
-                                         const int N,
-                                         const miopen::ConvolutionContext& params,
-                                         const bool fp16,
-                                         const unsigned filter_tile_size)
+namespace {
+/// \todo Consider re-using code from RxS_f2x3.
+inline bool IsShaderContraintsMet(const int R,
+                                  const int S,
+                                  const int,
+                                  const int,
```
Comment (atamazov) [inline src/solver/conv_winoRxS_f3x2.cpp:51]: Shall we remove unused params?

File: src/solver/conv_winoRxS_f3x2.cpp [hunk 2]
```diff
@@ -32,154 +32,42 @@
...
-        assert(padded_R % TILE == 0 && padded_S % (is_dilated_stride_2 ? TILE_X2 : TILE) == 0);
-        const int k = padded_R / TILE;
-        const int l = padded_S / (is_dilated_stride_2 ? TILE_X2 : TILE);
-        if(k * l * C < 18)
-            return false;
-    }
     // Padding for bwd data shall not be negative.
     /// \todo Either remove WrW related code or re-use function from RxS
-    if(params.direction.IsBackwardData() || params.direction.IsBackwardWrW())
+    if(params.direction.IsBackwardData())
```
Comment (atamazov) [inline src/solver/conv_winoRxS_f3x2.cpp:63]: Bugfix?

File: src/solver/conv_winoRxS_f3x2.cpp [hunk 3]
```diff
@@ -218,19 +105,13 @@ bool ConvBinWinogradRxSf3x2::IsApplicable(const ConvolutionContext& params) cons
...
         return false;
-    if(!(params.direction.IsForward() || params.direction.IsBackwardData()))
-        return false;
     if(!params.use_asm_kernels)
         return false;
-    if(!params.rmv.IsV2orV3())
-        return false;
-    if(!params.IsLayoutDefault())
-    {
+    if(!params.rmv.IsV3())
```
Comment (atamazov) [inline src/solver/conv_winoRxS_f3x2.cpp:110]: Do we support non-NCHW layout now?

File: src/solver/conv_winoRxS_f3x2.cpp [hunk 4]
```diff
@@ -239,141 +120,339 @@ bool ConvBinWinogradRxSf3x2::IsApplicable(const ConvolutionContext& params) cons
...
+    if(!IsWarned)
+    {
+        if(params.GetStream().GetMaxHardwareComputeUnits() > MAX_CU_LIMIT)
+            MIOPEN_LOG_WE(SolverDbId(*this) << ": GPU has "
+                                            << params.GetStream().GetMaxHardwareComputeUnits()
+                                            << "CUs, but this solver supports max "
+                                            << MAX_CU_LIMIT
+                                            << "and thus may show sub-optimal performance.");
+        IsWarned = true;
+    }
```
Comment (atamazov) [inline src/solver/conv_winoRxS_f3x2.cpp:176]: - Do you want to issue this message only once?
- Why would you like to issue an error message for DEV builds?
  - If this is an error, then maybe we shall make it non-applicable?

**Comment by atamazov** [comment]
@Slimakanzer Do you have any performance data? I would like to compare the performance of the new shader vs. the old one.

File: src/solver/conv_winoRxS_f3x2.cpp (hunk 1 above)
Comment (Slimakanzer) [reply src/solver/conv_winoRxS_f3x2.cpp]: Thank you! Fixed in 1e8b8a8

File: src/solver/conv_winoRxS_f3x2.cpp (hunk 2 above)
Comment (Slimakanzer) [reply src/solver/conv_winoRxS_f3x2.cpp:70]: This is an incorrect limitation that originally comes from `conv_winoRxS` solver.
Now wino_f3x2 has the same constraints as `conv_winoRxS_f2x3` solver, where this limitation was  [removed](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/640/commits/c20d531b086b179c5625ec1a642d1505991a8232).

File: src/solver/conv_winoRxS_f3x2.cpp (hunk 3 above)
Comment (Slimakanzer) [reply src/solver/conv_winoRxS_f3x2.cpp:115]: No, we dont. Constraint checked in the `IsShaderContraintsMet`: https://github.com/ROCmSoftwarePlatform/MIOpen/blob/bcf87b840cb07f728582bf745b9f76bf9565a220/src/solver/conv_winoRxS_f3x2.cpp#L71-L74

However, this check is not obvious, so it was fixed in e105526

File: src/solver/conv_winoRxS_f3x2.cpp (hunk 4 above)
Comment (Slimakanzer) [reply src/solver/conv_winoRxS_f3x2.cpp]: Shader still will work, but there might be severe performance penalties. It's better to set as non-applicable in this case.
Thank you, fixed in 9db3bc6

File: src/solver/conv_winoRxS_f3x2.cpp (hunk 3 above)
Comment (atamazov) [reply src/solver/conv_winoRxS_f3x2.cpp:115]: I see, thanks.

File: src/solver/conv_winoRxS_f3x2.cpp (hunk 4 above)
Comment (atamazov) [reply src/solver/conv_winoRxS_f3x2.cpp]: Good!

File: src/kernels/Conv_Winograd_v21_1_2_f3x2_fp32_stride1_group.s [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+ *******************************************************************************/
+.include "Conv_Winograd_v21_1_2_metadata.inc"
+
+KERNEL_PROLOG f3x2_fp32_stride1_group
+
+.if (.amdgcn.gfx_generation_number == 9)
+    .if (.amdgcn.gfx_generation_stepping == 10)
+        .error "gfx90a is not supported yet"
+    .else
+        .include "Conv_Winograd_v21_1_2_f3x2_gfx9_fp32_stride1_group.inc"
```
Comment (Kirpich30000) [inline src/kernels/Conv_Winograd_v21_1_2_f3x2_fp32_stride1_group.s:34]: Please make file naming consistent with kernel names. Code in conv_winoRxS_f3x2.cpp might also be simplified if the kernel name and file name would be consistent.

**Review by atamazov — DISMISSED** [review]
LGTM!

File: src/kernels/Conv_Winograd_v21_1_2_f3x2_fp32_stride1_group.s (hunk 5 above)
Comment (atamazov) [reply src/kernels/Conv_Winograd_v21_1_2_f3x2_fp32_stride1_group.s:34]: @Kirpich30000 You review does not block merging. Is it wnat you want, i.e. is it Ok to merge this without resolving your review comments first?

Comment (Kirpich30000) [reply src/kernels/Conv_Winograd_v21_1_2_f3x2_fp32_stride1_group.s:34]: I preffer to resolve the issues.

**Review by Kirpich30000 — CHANGES_REQUESTED** [review]
See comments

File: src/kernels/Conv_Winograd_v21_1_2_metadata.inc [hunk 6]
```diff
@@ -210,20 +210,28 @@ METADATA_WRAPPER total_sgpr_count,.amdgcn.next_free_vgpr,workgroup_size_x, <\ker
 
 .endm
 
-.macro PROLOG_KERNEL_DESCRIPTOR_WRAPPER machine_version, kernel_name_postfix
-	PROLOG_KERNEL_DESCRIPTOR miopenSp3AsmConv_v21_1_2_gfx\machine_version\()_\kernel_name_postfix
+.macro PROLOG_KERNEL_DESCRIPTOR_WRAPPER machine_version, kernel_name_postfix, kernel_name_prefix
```
[the remark this answers is not in the document]
Comment (Slimakanzer) [reply src/kernels/Conv_Winograd_v21_1_2_metadata.inc]: Thank you, fixed in 8af72ae

**Comment by Slimakanzer** [comment]
@atamazov
I apologize for the discarded reviews. Can you check new changes?

**Comment by atamazov** [comment]
@Slimakanzer No problem. I'll check but review comments are from @Kirpich30000 )))

**Review by Kirpich30000 — APPROVED** [review]
LGTM

**Pull request merged**